### PR TITLE
Add source_defined_primary_key validation test to test_read

### LINKED ISSUE #2748: SAT should verify that if PK is set in catalog it is never null for all output records
## Tell us about the problem you're trying to solve
Add testing in the _source_ standard tests to verify PK incremental functionality.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368157885506) by [Unito](https://www.unito.io)

### LINKED ISSUE #5142: SAT: Fix configured_catalog stream field fulfillment
## Enviroment
- **Airbyte version**:  0.29.0-alpha
- **Severity**: Medium

## Current Behavior
SAT reads `configured_catalog` from config file and does not fill it with python-specified fields. Though `discover` command add these fields to the catalog output.

## Expected Behavior
`configured_catalog` arg in SAT should be filled the same way, as the catalog generated by `discover`.

## Steps to Reproduce
1. Specify the `primary_key` in any stream
2. Does not specify `source_defined_primary_key` in configured catalog
3. Run `discover` command
4. Debug SAT to see the `configured_catalog` values
5. Compare the output from the steps 3 and 4

## Additional Info
This issue may be very important. 
Discover command [generates](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L76-#L79) catalog with 
```
streams = [stream.as_airbyte_stream() for stream in self.streams(config=config)]
return AirbyteCatalog(streams=streams)
```
Where the `as_airbyte_stream()` method actually fills the fields with python specified data.

So I believe that we may call discover command and get its results to then use in another tests (especially `test_read` where we need generated values, but it may be needed for other tests in future).

### PR DESCRIPTION
**Pull request by htrueman**
## What
Closes #2748

## How
Updated `TestBasicRead.test_read` test.
Added validation to check if primary key presents and not null in output records in case when `source_defined_primary_key` set in cofigured catalog.

## Recommended reading order
1. `test_core.py`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in output spec
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Documentation updated 
    - [ ] README
    - [ ] CHANGELOG.md
    - [ ] Reference docs in the `docs/integrations/` directory.
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 1]
```diff
@@ -105,6 +105,22 @@ def test_read(
 
         assert records, "At least one record should be read using provided catalog"
 
+        for stream in [stream for stream in configured_catalog.streams if stream.stream.source_defined_primary_key]:
+            stream_records = [r for r in records if r.stream == stream.stream.name]
+            for stream_record in stream_records:
+                for sub_pk in stream.stream.source_defined_primary_key:
```
Comment (Zirochkaa) [inline airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:111]: What is `sub_pk`? It's not clear from such name.

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 2]
```diff
@@ -105,6 +105,22 @@ def test_read(
...
+                    record_data = stream_record.data
+                    for sub_pk_part in sub_pk:
+                        if isinstance(record_data, dict):
+                            record_data = record_data.get(sub_pk_part)
+                        else:
+                            record_data = None
+
+                        assert record_data, (
+                            f"Primary key subkeys {repr(sub_pk)} "
+                            f"have null values or not present in {stream.stream.name} stream records. "
```
Comment (Zirochkaa) [inline airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:121]: There is an odd space here at the end of the line before `"`.

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 3]
```diff
@@ -105,6 +105,22 @@ def test_read(
 
         assert records, "At least one record should be read using provided catalog"
 
+        for stream in [stream for stream in configured_catalog.streams if stream.stream.source_defined_primary_key]:
+            stream_records = [r for r in records if r.stream == stream.stream.name]
+            for stream_record in stream_records:
+                for sub_pk in stream.stream.source_defined_primary_key:
+                    record_data = stream_record.data
+                    for sub_pk_part in sub_pk:
+                        if isinstance(record_data, dict):
```
Comment (Zirochkaa) [inline airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:114]: You can use ternary conditional operator here:
```
record_data = record_data.get(sub_pk_part) if isinstance(record_data, dict) else None
```
WDYT?

**Review by Zirochkaa — CHANGES_REQUESTED** [review]

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 4]
```diff
@@ -105,6 +105,22 @@ def test_read(
 
         assert records, "At least one record should be read using provided catalog"
 
+        for stream in [stream for stream in configured_catalog.streams if stream.stream.source_defined_primary_key]:
```
[the remark this answers is not in the document]
Comment (vitaliizazmic) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: Filtering can be implemented without list comprehension.

[the remark this answers is not in the document]
Comment (Zirochkaa) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: Do you mean using `filter()` function?

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py (hunk 2 above)
Comment (vitaliizazmic) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:121]: @htrueman is it necessary to show message that certain primary key missed? Or will be enough message, that some PK is missed?

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vitaliizazmic) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: Yes

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py (hunk 2 above)
Comment (htrueman) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:121]: I think showing message that certain primary key missed is much more convenient, so I've added this log on purpose.
If we have composed PK like this: `[["pk1", "pk2"], ["pk3", "pk4"]]`, then this log will show you which exact part of the PK is broken so you could debug it closely.

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (htrueman) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: `filter()` is not so pythonic and I can't see anything wrong with list comprehensions. Do you have any warnings about using it here?

[the remark this answers is not in the document]
Comment (Zirochkaa) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: I don't :)

**Review by Zirochkaa — APPROVED** [review]

**Review by vitaliizazmic — APPROVED** [review]

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 5]
```diff
@@ -105,6 +105,20 @@ def test_read(
 
         assert records, "At least one record should be read using provided catalog"
 
+        streams_with_primary_key = [stream for stream in configured_catalog.streams if stream.stream.source_defined_primary_key]
```
Comment (keu) [inline airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:108]: agreed to change

**Review by keu — APPROVED** [review]
lgmt with small changes that we discussed offline

**Pull request merged**